### PR TITLE
Single gate in moment

### PR DESCRIPTION
**Pull request by dretta**
Implements #3773

### CODE REVIEW

File: cirq-core/cirq/ops/gate_operation.py [hunk 1]
```diff
@@ -46,7 +46,7 @@
 
 @value.value_equality(approximate=True)
 class GateOperation(raw_types.Operation):
-    """An application of a gate to a sequence of qubits.
+    """An application of a gate to a sequence of qubits.raw_types.Operation
```
Comment (dstrain115) [inline cirq-core/cirq/ops/gate_operation.py:49]: What is this?  Typo?

File: cirq-core/cirq/ops/gate_operation.py [hunk 2]
```diff
@@ -334,5 +334,7 @@ def _equal_up_to_global_phase_(
             return False
         return protocols.equal_up_to_global_phase(self.gate, other.gate, atol=atol)
 
+    def equal_gates(self, other):
```
Comment (dstrain115) [inline cirq-core/cirq/ops/gate_operation.py:337]: Needs docstring. Does this really need to be a function on GateOperation?

File: cirq-core/cirq/ops/moment.py [hunk 3]
```diff
@@ -483,6 +483,26 @@ def _commutes_(
 
         return True
 
+    def get_single_gate_from_moment(self) -> Optional['cirq.Operation']:
```
Comment (dstrain115) [inline cirq-core/cirq/ops/moment.py:486]: Should this return a cirq.Gate instead?  If the gates match (but the qubits do not match, why does it pick the first operation?  Also, this is called get_single_gate_from_moment not from_single_op_from_moment

File: cirq-core/cirq/ops/moment_test.py [hunk 4]
```diff
@@ -617,3 +617,16 @@ def test_transform_qubits():
...
     with pytest.raises(TypeError, match='must be a function or dict'):
         _ = original.transform_qubits('bad arg')
+
+def test_single_gate_no_gates():
+    assert cirq.Moment() == None
+
+def test_single_gate_same_gates():
+    a = cirq.H(cirq.NamedQubit('a'))
+    b = cirq.H(cirq.NamedQubit('b'))
+    assert cirq.Moment([a, b]) == a
```
Comment (dstrain115) [inline cirq-core/cirq/ops/moment_test.py:627]: Do you mean 

 `assert cirq.Moment([a, b]).get_single_gate_from_moment()  == a `

File: cirq-core/cirq/ops/raw_types.py [hunk 5]
```diff
@@ -596,6 +599,8 @@ def _commutes_(
 
         return np.allclose(m12, m21, atol=atol)
 
+    def equal_gates(self, other):
```
Comment (dstrain115) [inline cirq-core/cirq/ops/raw_types.py:602]: Why is this here?  It looks like it could cause problems.

File: cirq-core/cirq/ops/gate_operation.py (hunk 2 above)
Comment (tanujkhattar) [reply cirq-core/cirq/ops/gate_operation.py:337]: This should definitely not be a function on `GateOperation` as the functionality below to check equality of underlying gates, should also work for other operation types like `TaggedOperations` etc. which won't happen as of now because the function only exists on GateOperations.

File: cirq-core/cirq/ops/moment.py [hunk 6]
```diff
@@ -483,6 +483,26 @@ def _commutes_(
...
+                it is repeated one or more times, the first operation is
+                returned.
+            None: When there is no operation or different operations found, or
+                the operation is not of 'cirq.GateOperation'.
+         """
+        if not self.operations:
+            return None
+
+        for op_index in range(len(self.operations) - 1):
+            if not self.operations[op_index].equal_gates(self.operations[op_index+1]):
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/moment.py:501]: You can the following checks to determine whether two operations have the same underlying gate. 
```python
def _have_equal_gates(op1, op2):
    if op1.gate is None or op2.gate is None:
        return False
    return op1.qubits == op2.qubits and op1.gate == op2.gate
```

This will make sure that other operations like `TaggedOperations`, `ControlledOperations`, `ParallelGateOperations` etc. work as expected when comparing the underlying gates.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]

**Comment by tanujkhattar** [comment]
Also, another PR is open for the same issue https://github.com/quantumlib/Cirq/pull/3930

We should make sure to merge only one of the two open PRs!

File: cirq-core/cirq/ops/moment.py (hunk 6 above)
Comment (dretta) [reply cirq-core/cirq/ops/moment.py:501]: Why compare qubits?

Comment (tanujkhattar) [reply cirq-core/cirq/ops/moment.py:501]: Oops, my bad. That's a typo. We should be comparing number of qubits, not the actual qubits. So. 
```python
def _have_equal_gates(op1, op2):
    if op1.gate is None or op2.gate is None:
        return False
    return len(op1.qubits) == len(op2.qubits) and op1.gate == op2.gate
```

This is required as of now, because of operations like `ParallelGateOperation`, which is an operation that applies an underlying single qubit gate on multiple qubits. I've also filed  #4358 to track that such an inconsistency shouldn't exist.

File: cirq-core/cirq/ops/gate_operation.py (hunk 1 above)
Comment (dretta) [reply cirq-core/cirq/ops/gate_operation.py]: Removed.

File: cirq-core/cirq/ops/gate_operation.py (hunk 2 above)
Comment (dretta) [reply cirq-core/cirq/ops/gate_operation.py]: Removed.

File: cirq-core/cirq/ops/moment_test.py (hunk 4 above)
Comment (dretta) [reply cirq-core/cirq/ops/moment_test.py]: Fixed.

File: cirq-core/cirq/ops/raw_types.py (hunk 5 above)
Comment (dretta) [reply cirq-core/cirq/ops/raw_types.py]: Removed.

File: cirq-core/cirq/ops/moment.py (hunk 6 above)
Comment (dretta) [reply cirq-core/cirq/ops/moment.py]: Implemented.

File: cirq-core/cirq/ops/moment.py (hunk 3 above)
Comment (dretta) [reply cirq-core/cirq/ops/moment.py]: Changed from Operation to Gate.

**Comment by dretta** [comment]
All issues seem to be resolved.

**Comment by 95-martin-orion** [comment]
> Also, another PR is open for the same issue #3930

I've been in touch with the author of that PR, they're very busy currently and OK with this PR taking precedence.

File: cirq-core/cirq/ops/moment.py [hunk 7]
```diff
@@ -49,6 +49,18 @@ def _default_breakdown(qid: 'cirq.Qid') -> Tuple[Any, Any]:
         return None, qid
 
 
+def _have_equal_gates(op1, op2):
```
Comment (dstrain115) [inline cirq-core/cirq/ops/moment.py:52]: Add type information, ie.

`def _have_equal_gates(op1: cirq.Operation, op2: cirq.Operation)->bool:`

File: cirq-core/cirq/ops/moment.py [hunk 8]
```diff
@@ -49,6 +49,18 @@ def _default_breakdown(qid: 'cirq.Qid') -> Tuple[Any, Any]:
...
+    """Compares two operations and returns a boolean of it their gates
```
Comment (dstrain115) [inline cirq-core/cirq/ops/moment.py:53]: Prefer a single line summary at the beginning.  Can be changed to:

"""Compares two operations and returns True if their gates match."""

File: cirq-core/cirq/ops/moment.py [hunk 9]
```diff
@@ -49,6 +49,18 @@ def _default_breakdown(qid: 'cirq.Qid') -> Tuple[Any, Any]:
...
+    match.
+    Returns:
+        bool: True if both gates are not null, the same, and take in
+           the same number of qubits from their operations.
+    """
+    if op1.gate is None or op2.gate is None:
+        return False
+    return len(op1.qubits) == len(op2.qubits) and op1.gate == op2.gate
```
Comment (dstrain115) [inline cirq-core/cirq/ops/moment.py:61]: Why does the number of qubits matter?  If I have cirq.measure(a), cirq.measure(b), cirq.measure(c) that returns MesaurementGate, but cirq.measure(a,b), cirq.measure(c) returns None?  That's unexpected.

File: cirq-core/cirq/ops/moment_test.py [hunk 10]
```diff
@@ -617,3 +617,20 @@ def test_transform_qubits():
...
+
+
+def test_single_gate_same_gates():
+    a = cirq.H(cirq.NamedQubit('a'))
+    b = cirq.H(cirq.NamedQubit('b'))
+    c = cirq.H(cirq.NamedQubit('c'))
+    assert cirq.Moment([a, b, c]).get_single_gate_from_moment() == cirq.H
+
+
+def test_single_gate_different_gates():
```
Comment (dstrain115) [inline cirq-core/cirq/ops/moment_test.py:633]: Should we add a test for different numbers of qubits (if we keep that behavior)?

File: cirq-core/cirq/ops/moment.py (hunk 7 above)
Comment (dretta) [reply cirq-core/cirq/ops/moment.py]: Done.

File: cirq-core/cirq/ops/moment.py (hunk 8 above)
Comment (dretta) [reply cirq-core/cirq/ops/moment.py]: Done.

File: cirq-core/cirq/ops/moment.py (hunk 9 above)
Comment (dretta) [reply cirq-core/cirq/ops/moment.py]: Done.

File: cirq-core/cirq/ops/moment_test.py (hunk 10 above)
Comment (dretta) [reply cirq-core/cirq/ops/moment_test.py:633]: Done.

**Comment by dretta** [comment]
All issues seem to be resolved.

File: cirq-core/cirq/ops/moment.py [hunk 11]
```diff
@@ -50,6 +50,17 @@ def _default_breakdown(qid: 'cirq.Qid') -> Tuple[Any, Any]:
...
 
+def _have_equal_gates(op1: 'cirq.Operation', op2: 'cirq.Operation') -> bool:
+    """Compares two operations and returns True if their gates match.
+    Returns:
+        bool: True if both gates are not null, the same, and take in
+           the same number of qubits from their operations.
+    """
+    if op1.gate is None or op2.gate is None:
+        return False
+    return isinstance(op1.gate, type(op2.gate)) and isinstance(op2.gate, type(op1.gate))
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/moment.py:61]: We should compare the gate instances and not have an `isinstance` check. Gate instances with different parameter values will correspond to different gates. 

Eg: `cirq.X ** 0.25` and `cirq.X ** 0.5` are two different gates but your function will return true. 

We can now simply return `op1.gate == op2.gate` as `ParallelGateOperation` is now deprecated, so there's no additional need to compare qubit length (which I had suggested in my previous review).

Comment (dretta) [reply cirq-core/cirq/ops/moment.py:61]: I would like to hear from @dstrain115 as well, seeing as he was concerned with comparing measurement gates and suggested not to use `op1.gate == op2.gate`.

Comment (dstrain115) [reply cirq-core/cirq/ops/moment.py:61]: Did I recommend isinstance?  I don't remember why and can't find the comment.  My previous comment was regarding comparing the number of qubits, which would fail for gates that can have variable number of qubits (such as measure gates).

Tanuj is correct though, this check will fail for gates with different parameters.

File: cirq-core/cirq/ops/moment_test.py [hunk 12]
```diff
@@ -617,3 +617,41 @@ def test_transform_qubits():
...
+    a = cirq.H(cirq.NamedQubit('a'))
+    b = cirq.X(cirq.NamedQubit('b'))
+    assert cirq.Moment([a, b]).get_single_gate_from_moment() is None
```
Comment (dstrain115) [inline cirq-core/cirq/ops/moment_test.py:636]: Maybe add this test to see the problem clearly:

```
    a = cirq.XPowGate(exponent=0.25)(cirq.NamedQubit('a'))
    b =cirq.XPowGate(exponent=0.5)(cirq.NamedQubit('b'))
    assert cirq.Moment([a, b]).get_single_gate_from_moment() is None
```

Comment (dretta) [reply cirq-core/cirq/ops/moment_test.py:636]: Done.

File: cirq-core/cirq/ops/moment.py (hunk 11 above)
Comment (dretta) [reply cirq-core/cirq/ops/moment.py]: Done.

**Comment by dabacon** [comment]
@dretta is this still in flight or should we assign the issue to someone else?  Just checking as this has been open for a while now.

**Comment by dretta** [comment]
I thought that the team didn't want this feature anymore, is this still the case?

https://docs.google.com/document/d/1K3e4MtNsWBlQ4KWkpJftac9AgBjHygvWqtucBdRoZyQ/edit#bookmark=id.s0bl0turvt0d

**Comment by tanujkhattar** [comment]
From cirq sync notes on November 3, 2021:
"Decision: Let’s abandon this PR since there are too many issues with the requirements. Sorry Daniel!"

I'll close this PR, apologies for the confusion Daniel.
